src/main/java/hudson/tasks/junit/TestResult.java

```diff
@@ -306,14 +310,20 @@ public void parse(File reportFile) throws IOException {
         } catch (RuntimeException e) {
             throw new IOException("Failed to read "+reportFile,e);
         } catch (DocumentException e) {
+            String reportFileContent = null;
+            try {
+                reportFileContent = Files.toString(reportFile, UTF_8);
+            } catch (Exception ee) {
+                /* No need to further explain the error if we end up here. */
+            }
             if (!reportFile.getPath().endsWith(".xml")) {
-                throw new IOException("Failed to read "+reportFile+"\n"+
-                    "Is this really a JUnit report file? Your configuration must be matching too many files",e);
+                throw new IOException("Failed to parse "+reportFile+"\n"+
+                    "Is this really a JUnit report file? Your configuration may be matching too many files? The file contains:\n\n" + reportFileContent + "\n\n",e);
```

Comment:
> We reject it because it cannot be parsed and suggest that it may not be a report file because of its name.

Right, I missed that.

> I don't see a problem with having a very large log in those rare cases.

This is not about build logs (only). You are reading potentially megabytes of data into string eagerly, transferring it over the network and reconstructing it on master side. Putting aside the memory pressure, this is an effective way to dos the system either intentionally or accidentally.

Trimming the file to, say, 1KB would help user to identify non-result files are included but filename would almost just as well IMO. In case of "real" parsing problem trimming can omit the offending part.

The best alternative would be to get SAX report more useful error so we do not need to forward the file to the user at all.

If we absolutely must report the file as a whole, it should be transferred as a file (so compressed and buffered without loading it whole into the memory).

@jenkinsci/code-reviewers, thoughts?